### PR TITLE
Ducarroz/remote webpack config

### PR DESCRIPTION
**Pull request by ducarroz**
:bird: @zendesk/harrier Instead of patching the webpack config directly into react-script, add a hook that allow the consumer to define a method to customize the config. In lotus, you now just need to add the following `zendesk_console/lotus_react/config/webpack.config.patch.js`:

```
'use strict';

module.exports = (config, environment, paths) => {
  // Allow absolute path (rooted to the src folder)
  config.resolve.modules.push(paths.appSrc);

  // Patch loaders
  config.module.rules.forEach(rule => {
    if (rule.loader && rule.loader.indexOf('/file-loader/') !== -1) {
      if (rule.exclude && rule.exclude.join(',').indexOf('/\\.svg$/') === -1) {
        // add support for SVG files
        rule.exclude.push(/\.svg$/);
      }
      // ???
      rule.options.name = 'agent/assets/react/media/[name].[hash:8].[ext]';
    }

    else if (rule.loader && rule.loader.indexOf('/url-loader/') !== -1) {
      if (rule.test && rule.test.join(',').indexOf('/\\.svg$/') === -1) {
        // add support for SVG URLs
        rule.test.push(/\.svg$/);
      }
      // ???
      rule.options.name = 'agent/assets/react/media/[name].[hash:8].[ext]';
    }

    else if (rule.test.toString() === '/\\.css$/') {
      const loader = environment === 'prod' ? rule.loader : rule.use;

      loader.forEach(rule => {
        if (rule.loader && rule.loader.indexOf('/css-loader/') !== -1) {
          // enable css module
          rule.options.modules = true;
          rule.options.localIdentName = '[name]__[local]___[hash:base64:5]';
        }
      });
    }
  });

  return config;
};
```

/cc @bmson 

## References
- JIRA: n/a

## Risks
- Low

### CODE REVIEW

**Comment by bmson** [comment]
@zendesk/orchid

**Comment by rapheld** [comment]
Where is the `webpack.config.patch.js` file?

**Comment by shajith** [comment]
I wonder if we should try to submit a patch like this to upstream create-react-app. Maybe there's an established precedent of the maintainers not interested in this sort of customization?

Also, subjectively, I like "override" as a better name than "patch", from the perspective of proposing as a change to create-react-app.

**Comment by ducarroz** [comment]
the `webpack.config.patch.js` is provided by the consumer project. In our case, lotus_react

**Comment by bmson** [comment]
Nothing against this idea. Just curious how it will behave when we eject... If that all works I'm +1

Also worth considering if it's worth having separate dev and prod overwrite?

**Comment by ducarroz** [comment]
At first I was going to have separate override, but as both files were practically identical, I decided to use only one file. As you receive the environment (mode), nothing prevent you to split it into separate file on the consumer side if needed by having `zendesk_console/lotus_react/config/webpack.config.patch.js` loading separate dependencies

Regarding using the term "override" rather than "path", I totally agree, I will change it...

**Comment by shajith** [comment]
I like this approach. The officially supported approach to customize CRA seems to be forking react-scripts, and this lets us customize options via a overriding config file, when using our fork. This way, we can just edit the overrides when we want to tweak config, and only need to update and publish our fork when we want to sync up with upstream CRA.

Is there any kind of customization that's impossible via a webpack config override?

Tangential: webpack config extension has some library support, eg: https://github.com/Fitbit/webpack-config

File: packages/react-scripts/config/webpack.config.dev.js [hunk 1]
```diff
@@ -294,3 +291,14 @@ module.exports = {
...
 };
+
+const configPath = path.resolve(paths.appPath, 'config', 'webpack.config.patch.js');
+let customizeConfig = null;
+
+try {
+  customizeConfig = require(configPath);
+} catch (ex) {
+  console.error("No custom config file found:", configPath.slice(paths.appPath.length + 1));
+}
```
Comment (shajith) [inline packages/react-scripts/config/webpack.config.dev.js]: maybe glance at https://github.com/Fitbit/webpack-config for what they do?

File: packages/react-scripts/config/webpack.config.dev.js [hunk 2]
```diff
@@ -294,3 +291,14 @@ module.exports = {
     hints: false,
   },
 };
+
+const configPath = path.resolve(paths.appPath, 'config', 'webpack.config.patch.js');
+let customizeConfig = null;
+
+try {
+  customizeConfig = require(configPath);
+} catch (ex) {
```
Comment (shajith) [inline packages/react-scripts/config/webpack.config.dev.js]: Why not just check if the file exists, and then require it? or just silently move on in the absence of a file? not having an override file isn't an error, right?

(btw: thinking through how we can make our fork generally useful for other CRA users in Zendesk)

Comment (ducarroz) [inline packages/react-scripts/config/webpack.config.dev.js]: yep, I was too lazy that day :-), will change that too...

File: packages/react-scripts/config/webpack.config.dev.js (hunk 1 above)
Comment (ducarroz) [inline packages/react-scripts/config/webpack.config.dev.js]: looking...

Comment (ducarroz) [inline packages/react-scripts/config/webpack.config.dev.js]: I took a look at https://github.com/Fitbit/webpack-config and while this is an interesting generic approach, it wont really help us in our case. The main reason is that we need to be able to add some value (file type) to existing array which Fitbit/webpack-config wont do. We would have to override the whole array instead which I don't like, would introduce more work and dependencies on our side, at least for now.

**Comment by ducarroz** [comment]
bump! good for review now...

**Comment by shajith** [comment]
@ducarroz: do you plan to convert the existing mods to create-react-app to a custom override file? (the changes we already made, I mean, to support CSS-modules, and some other things)

**Comment by ducarroz** [comment]
@shajith correct. But before I can take care of the lotus side, I need to publish this PR

**Comment by shajith** [comment]
cc: @zendesk/orchid

**Pull request merged**